### DISCUSSION TITLE
Is there a way to have two folders of the repo? One Running Torch 2 and another running old Torch?

### DISCUSSION THREAD

**Discussion by spiritparticle** [vladmandic/sdnext#1645] [Q&A]
This might sound crazy. But I have a bunch of really old images that I just LOVE and really want to generate more of but ever since the update to torch v2 I cannot get the same kinds of images. And just wondering if there is a way or how I could run two versions. One to stay up to date and contemporary with all the updates. And one that is old and uses old torch and I can use that when I want to generate some of my "vintage" images.

**Comment by vladmandic** [comment]
in theory, yes, since each install uses its own venv. in practice, you'll have to try :)

**Reply by spiritparticle** [reply]
Is it possible... would it even be a good idea? To maybe ask or request for a way or an option to toggle between torch versions? Or would that just screw with everything and all the work you have done?

**Reply by vladmandic** [reply]
you can already use `TORCH_COMMAND` env variable to force specific torch version. switch on the fly? no. separate installs like you mentioned? likely possible.

**Reply by spiritparticle** [reply]
That sounds very promising! I am sorry to ask since I am very very new to understanding code. I literally only just learned what a venv is and how to kind of make one and the basics thanks to chatGPT but... How do I use that command? Or what do I do to use that command or hmm how do I ask this... do I change the variable some place? I am not sure what I am asking... Umm. How do I do the thing that you are suggesting. LOL

**Reply by spiritparticle** [reply]
Okay so I have found two references to TORCH_COMMAND in the installer.py file. There are two references to the version number and torchvision. Do I simply make a new folder git pull a fresh copy of the repo and change those version numbers to something like pytorch==1.13.1 torchvision==0.14.1 and save and run the webui and let it all install that way? I am honestly kind of proud that I managed to even get this far lol.

EDIT: No that didn't work. I might have to try and understand this tomorrow. I am kind of confused and stuck at the moment.

**Reply by vladmandic** [reply]
re: TORCH_COMMAND is something you set as environment variable in your shell and then installer will pick it up if its set and use as override. if its NOT set, then it will use default.

re: venv - that's automatically created by webui.bat/webui.sh - thats pretty much the only reason they exist and then they run webui.py

**Reply by spiritparticle** [reply]
Could you please explain the steps I take to implement the override? I am understanding the concept. But I am not sure how to do what you are saying by creating a variable override in the shell. If you could explain to me what to do I would appreciate it so much. I want to thank you also for your help so far.

**Reply by brknsoul** [reply]
You'd make two batch files; 

`oldtorch.bat`
```
SET TORCH_COMMAND=<old torch variable here>
webui.bat %*
```

`newtorch.bat`
```
SET TORCH_COMMAND=<new torch variable here>
webui.bat %*
```

You'll have to find out what the variables you need are.
The `%*` is there so you can pass command line args to webui.bat, like `--medvram`, `--autolaunch`, and so on.

**Reply by spiritparticle** [reply]
Thankyou so much I am going to try it right now. Just to make extra sure I am doing it correctly I will show what I THINK is the right thing to do and I am also going to try also guessing an example command line arg too like xformers.

`SET TORCH_COMMAND=pytorch==1.13.1 torchvision==0.14.1
webui.bat %*--xformers`

Is that right?

EDIT: It seems to be working! But I had to change it to the old linux install line "torch==1.13.1+cu116 torchvision==0.14.1+cu116 torchaudio==0.13.1 --extra-index-url https://download.pytorch.org/whl/cu116"

Just another quick question. Could I use this method to change other things also? Like even load the old transformers to make REALLY REALLY old images?

**Reply by brknsoul** [reply]
I don't think so.. you'll have to wait until some with more 🧠comes along ;-)

**Reply by spiritparticle** [reply]
Well thankyou so much anyway for your help. The repos are still being cloned and everything is still being installed but I THINK this is working so thank you again!

Usually I just ran a very very old commit of auto1111 repo but they just don't seem to work anymore.

EDIT: Actually this is strange... the --xformers arg is unrecognized. Do you have any ideas why that might be? It lists all the args and i see --medvram and --lowvram and a bunch of others but no xformers arg

**Reply by vladmandic** [reply]
In sdnext it's not cmd arg, it's in settings, like most of others.

**Reply by spiritparticle** [reply]
Thanks vladmandic. I was literally reading a development update in another tab and was about to come back and correct myself but saw your reply. I will try to read up and research a bit more before blindly asking questions again.

**Reply by spiritparticle** [reply]
I am very sorry. But I can find the optimization method in the settings but I am not sure how to install xformers. I would like to use xFormers but use the Doggettx's cross attention method.

**Reply by brknsoul** [reply]
![image](https://github.com/vladmandic/automatic/assets/1969381/9030d85d-2696-47ea-85c9-52e17bd0c289)

both xformers and doggettx is a cross attention method, you use one or the other.
Typically you use xformers for a low-end nvidia GPU, or Scale Dot Product for higher end GPUs
